speechbrain/utils/distributed.py

```diff
@@ -196,10 +196,9 @@ def ddp_barrier():
     if MAIN_PROC_ONLY >= 1 or not is_distributed_initialized():
         return
 
-    if torch.distributed.get_backend() == "nccl":
+    if torch.distributed.get_backend() == torch.distributed.Backend.NCCL:
         # if NCCL, we can retrieve device_ids through this way
-        device_ids = [int(os.environ.get("LOCAL_RANK"))]
-        torch.distributed.barrier(device_ids=device_ids)
+        torch.distributed.barrier(device_ids=[torch.cuda.current_device()])
```

Comment:
This doesn't seem right to me, shouldn't the device_ids contain all the devices in the group? As is it seems like it wouldn't block at all

Reply:
I agree that it seems a bit "weird". I would invite you to read about this conversation by the PyTorch Lightning team: https://github.com/Lightning-AI/pytorch-lightning/pull/8165 and see code from multiple projects (https://github.com/Lightning-AI/pytorch-lightning/blob/7d2b85916dcefd91df92edcb69411e97e9ac4a65/src/lightning/pytorch/strategies/ddp.py#L291-L299 ; https://github.com/pytorch/tnt/pull/397 ).

I think what is happening when you do that, is that you are for forcing the barrier to be exactly on this process (see: https://github.com/pytorch/pytorch/issues/53658) which might not be the case.

Reply:
BTW, i really confirm that if you are not setting the cuda device, `torch.cuda.current_device()` will always return `0`. And this makes a difference when using torch.distributed.barrier. You really need to make sure that it returns the same value as `LOCAL_RANK` otherwise the barrier wont work properly.

Reply:
I might have found the exact explanation of the issue: 

![Capture d’écran 2024-09-15 à 21 09 12](https://github.com/user-attachments/assets/a5859a47-8e3a-4a31-84e3-76b77e2d5182)
Source: https://pytorch.org/docs/stable/distributed.html#launch-utility

Reply:
BTW, simplifying `ddp_barrier` to:
```
def ddp_barrier():
    r"""
    Synchronize all processes in distributed data parallel (DDP) mode.

    This function blocks the execution of the current process until all
    processes in the distributed group have reached the same point. It ensures
    that no process moves ahead until every other process has also reached this
    barrier. If DDP is not being used (i.e., only one process is running),
    this function has no effect and immediately returns.

    Returns
    -------
    None


    Example
    -------
    >>> ddp_barrier()
    >>> print("hello world")
    hello world
    """
    if MAIN_PROC_ONLY >= 1 or not is_distributed_initialized():
        return

    torch.distributed.barrier()
```

Does also work! Adding `device_ids` is not what is making `torch.distributed.barrier` work. This is when we are setting the cuda device. I believe having `device_ids` is solving some issue that you can have when you are only using e.g. gpu 1 and 3 from a node with more GPUs.

Reply:
So the PyTorch docs are inconsistent about the use of this function. Love it.